docs/code/introduction/machine_code.md

```diff
@@ -0,0 +1,257 @@
+
+MPF contains a “CustomCode” concept which lets you add custom code to your game.
+
+CustomCode classes are Python modules that run at the “root” of your game. You can use them to do anything you want.
+
+Note that MPF also has the ability to run custom mode code which is code that is associated with a certain game mode and is generally only active when the mode it’s in is active. So if you just want to write your own custom game logic, you’ll probably use mode code.
+
+CustomCode classes, on the other hand, are sort of “machine-level” custom code. CustomCode classes are nice if you have some kind of custom device type that doesn’t match up to any of MPF’s built in devices. The elevator and claw unloader in Demolition Man is a good example, and what we’ll use here.
+
+(You can read about how to download and run Demo Man in the example games section section of the MPF User Documentation.)
+
+Here’s how to create a custom code class:
+## Create your custom code file
+
+First, add a `custom_code` folder to your machine folder (you can use another name if you want). Then inside there, create the Python file that will hold your custom code classes. You can name this file whatever you want, just remember the name for the next step.
+
+Add an empty `__init__.py` file into your folder to make it a package. It become the package code and all your classes will be referenced as `custom_code.file_name.ClassName`.
+
+## Open and edit your custom code class file
+
+Next, edit the class file you created. At a bare minimum, you’ll need this:
+
+```
+from mpf.core.custom_code import CustomCode
+
+class Claw(CustomCode):
+    pass
+```
+
+Note that MPF contains a CustomCode base class which is very simple. You can see the source of it on GitHub [here](https://github.com/missionpinball/mpf/blob/dev/mpf/core/custom_code.py). We called our class Claw in this case.
+
+Pretty much all this does is give you a reference to the main MPF machine controller at self.machine, as well as setup a delay manager you can use and set the name of your class. There’s also an on_load() method which is called when the class is loaded which you can use in your own code.
+
+## Add the class to your machine config
+
+Next, edit your machine config file and add a `custom_code:` section, then under there add the package (folder), followed by a dot, then the module (file name) for your class, followed by a dot, followed by the class name for your class.
+
+For Demo Man, that looks like this:
+
+```
+custom_code:
+  - code.claw.Claw
+```
+
+This references class Claw in file claw.py which lives package custom_code.
+
+## Real-world example
+
+At this point you should be able to run your game, though nothing should happen because you haven’t added any code to your code.
+
+Take a look at the final Demo Man claw class to see what we did there. Since custom code classes have access to self.machine and they load when MPF loads, you can do anything you want in them.
+
+
```

Comment:
This is a big block of code :/

Is this file on GitHub somewhere that we could just link to?

Reply:
No idea, it came from the original docs. If there is a place where al configs are stored (for example for automated testing and validation), then let me know where that place is and I can do that. I believe in one post there was mentioned something like that some time ago, but cannot remember details.